caput/tools.py

```diff
@@ -66,3 +66,71 @@ def unique_ordered(x: Iterable) -> list:
     seen_add = seen.add
 
     return [i for i in x if not (i in seen or seen_add(i))]
+
+
+def allequal(obj1, obj2):
+    """Check if two objects are equal.
+
+    This comparison can check standard python types and numpy types,
+    even in the case where they are nested (ex: a dict with a numpy
+    array as a value).
+
+    Parameters
+    ----------
+    obj1 : scalar, list, tuple, dict, or np.ndarray
+    obj2 : scalar, list, tuple, dict, or np.ndarray
+    """
+    try:
+        _assert_equal(obj1, obj2)
+    except AssertionError:
+        return False
+    return True
+
+
+def _assert_equal(obj1, obj2):
+    """Assert two objects are equal.
+
+    This comparison can check standard python types and numpy types,
+    even in the case where they are nested (ex: a dict with a numpy
+    array as a value).
+
+    For more information:
+    https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_equal.html
+
+    Parameters
+    ----------
+    obj1 : scalar, list, tuple, dict, or np.ndarray
+    obj2 : scalar, list, tuple, dict, or np.ndarray
+    """
+    __tracebackhide__ = True
+    if isinstance(obj1, dict):
+        # Check that the dict-type objects are equivalent
+        if not isinstance(obj2, dict):
+            raise AssertionError(repr(type(obj2)))
+        _assert_equal(len(obj1), len(obj2))
+        # Check that each key:value pair are equal
+        for k in obj1.keys():
+            if k not in obj2:
+                raise AssertionError(repr(k))
+            _assert_equal(obj1[k], obj2[k])
+
+        return
+
+    if isinstance(obj1, (list, tuple)) and isinstance(obj2, (list, tuple)):
+        # Check that the sequence-type objects are equivalent
+        _assert_equal(len(obj1), len(obj2))
+        # Check that each item is the same
+        for k in range(len(obj1)):
+            _assert_equal(obj1[k], obj2[k])
+
+        return
+
+    # Ensure that any np.ndarray subclass is an np.ndarray
+    if isinstance(obj1, np.ndarray):
+        obj1 = obj1.view(np.ndarray)
+
+    if isinstance(obj2, np.ndarray):
+        obj2 = obj2.view(np.ndarray)
```

Comment:
What do you want to happen if someone does this?
```python
arr1 = mpiarray.MPIArray((4,) axis=0, comm=MPI.COMM_SELF)
arr1.local_array[:] = np.arange(4)

arr2 = np.arange(4)

tools._assert_equal(arr1, arr2)
```
It seems like you want them to evaluate to equal, but I'm not sure if that should generally be considered to be true.

Reply:
Ah I see you even wrote a test to cover this one. I'm still not sure if this is generally desirable though.

Reply:
I see the issue with having that compare to be equal. I'll tweak it a bit. 

Annoyingly, the only reason we have to take the view in the first place is because `numpy.testing.assert_equal` calls `.ravel` on `ndarray` subclasses, which obviously breaks with a `MPIArray`.